contrib/db_pools/lib/src/pool.rs

```diff
@@ -0,0 +1,235 @@
+use rocket::async_trait;
+
+/// This trait is implemented on connection pool types that can be used with the
+/// [`Database`] derive macro.
+///
+/// `Pool` determines how the connection pool is initialized from configuration,
+/// such as a connection string and optional pool size, along with the returned
+/// `Connection` type.
+///
+/// Implementations of this trait should use `async_trait`.
+///
+/// ## Example
+///
+/// ```
+/// #[derive(Debug)]
+/// struct Error { /* ... */ }
+/// # impl std::fmt::Display for Error {
+/// #     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+/// #         unimplemented!("example")
+/// #     }
+/// # }
+/// # impl std::error::Error for Error { }
+///
+/// struct Pool { /* ... */ }
+/// struct Connection { /* .. */ }
+///
+/// #[rocket::async_trait]
+/// impl rocket_db_pools::Pool for Pool {
+///     type Connection = Connection;
+///     type Config = rocket_db_pools::Config;
+///     type InitError = Error;
+///     type GetError = Error;
+///
+///     async fn initialize(config: Self::Config) -> Result<Self, Self::InitError> {
+///         unimplemented!("example")
+///     }
+///
+///     async fn get(&self) -> Result<Connection, Self::GetError> {
+///         unimplemented!("example")
+///     }
+/// }
+/// ```
+///
+/// [`Database`]: crate::Database
+#[async_trait]
+pub trait Pool: Sized + Send + Sync + 'static {
+    /// The type returned by get().
+    type Connection;
+
+    /// The configuration type this database pool expects.
+    type Config: rocket::serde::DeserializeOwned + Send;
+
+    /// The error type returned by `initialize`.
+    type InitError: std::error::Error;
+
+    /// The error type returned by `get`.
+    type GetError: std::error::Error;
```

Comment:
This feels wrong to me. Why isn't there just one `Error`?

Reply:
`InitError` is `rocket_db_pools::Error`, which represents either bad configuration or the database-specific error. `GetError` could be the same error type, but it would always be the non-configuration variant.

Reply:
For some database drivers (e.g. `deadpool-postgres`), the error type returned by initialization vs by `get()` are actually different. In that case one error type is a superset of the other, but I wouldn't want to assume that about future drivers.

Reply:
Sure, but that doesn't mean the pool integration can't wrap those into one type.

I thin there's a lot of effort right now spent on trying to creates these reusable structures like `Config` and `Error` when, in reality, each database driver (at least those from each library) will be pretty different. I would advocate for eschewing both of these types entirely and instead creating an `Error` and `Config` for each database variant/library as needed. With `Config` specifically, this is pretty important.

Reply:
> when, in reality, each database driver (at least those from each library) will be fairly distinct

So, I tried using a separate error type for each implementation and ran into every implementation of `Pool::initialize` needing to return a `figment::Error`. I feel again like I'm completely misunderstanding something, because the single `Error` type with 3 variants now seems like a lot less effort than writing and documenting the same 3 variants of a separate error enum for each database.